### DISCUSSION TITLE
How to Impose Reciprocity Constraint on Green's Function Calculation

### DISCUSSION THREAD

**Discussion by msomeya1** [lululxvi/deepxde#1454] [Q&A]
Dear Professor Lu Lu and DeepXDE Community,

First, I would like to express my gratitude for developing this incredible library. It has been very helpful in solving my problems.

Currently, I am interested in computing Green's function using PINNs. As an example, let us consider the 1D Poisson equation `-d^2G(x,x')/dx^2 = δ(x-x')` in `0<x<1`, with Dirichlet boundary condition (`G=0` at `x=0,1`). The following code snippet works correctly. (Note: I have substituted the Dirac delta function with a sharp Gaussian function to avoid singularity.)

```python
import deepxde as dde
import numpy as np
from deepxde.backend import tf

a = 0.02
Amp = 1.0

def Gaussian_tf(x, a):
    return tf.math.exp(-0.5*x**2/a**2) / (np.sqrt(2.0*np.pi) * a)

def pde(x, y):
    y_xx = dde.grad.hessian(y, x, i=0, j=0)
    f = Amp * Gaussian_tf(x[:, 0:1]-x[:, 1:2], a)
    return y_xx + f

def boundary(_, on_boundary):
    return on_boundary

geom = dde.geometry.Rectangle([0, 0], [1, 1])
bc = dde.icbc.DirichletBC(geom, lambda x: 0, boundary)

data = dde.data.PDE(geom, pde, bc, num_domain=1200, num_boundary=120, num_test=1500)
net = dde.nn.FNN([2] + [50] * 4 + [1], "tanh", "Glorot uniform")
model = dde.Model(data, net)

model.compile("adam", lr=0.001)

losshistory, train_state = model.train(iterations=10000)
dde.saveplot(losshistory, train_state, issave=True, isplot=True)
```

I want to implement further constraint, i.e., the reciprocity `G(x,x') = G(x',x)`. Could you  provide guidance on how I can incorporate this reciprocity constraint into the existing DeepXDE framework? I found some features that might be useful (e.g., `PointSetOperatorBC`), but I don't know if that will help. Perhaps additional custom losses need to be defined?

I would greatly appreciate any insights you could provide to help address this issue. Since I have recently started working with Python and machine learning, and this is my first time sending a question on GitHub, I hope I am not making a mistake or being disrespectful in any way. Thank you in advance.

[PS]
I found this issue https://github.com/lululxvi/deepxde/issues/631, but it is about one-variable symmetry, so I don't know if it is useful.

**Comment by msomeya1** [comment]
Dear Professor Lu Lu and DeepXDE Community,

I have found some mistakes in my previous code and made corrections. Sorry for the confusion.

By the way, I came up with one idea to satisfy the reciprocity; split the `pde` into two equations as follows:
```python
def pde(x, y):
    y_xx = dde.grad.hessian(y, x, i=0, j=0)
    y_xixi = dde.grad.hessian(y, x, i=1, j=1)
    f = Amp * Gaussian_tf(x[:, 0:1]-x[:, 1:2], a)
    return [y_xx + f, y_xixi + f]
```
I have conducted numerical experiments locally, and it seems to be working well. 

However, I'm still curious if there's another approach. 
For instance, is it possible to define a loss function to satisfy reciprocity, which includes term like `G(x1, x2) - G(x2, x1)`, where `x1` and `x2` are the array of training points of two coordinates.

Thank you in advance.
